src/consensus/read_validator.rs

```diff
@@ -98,8 +104,74 @@ impl ReadValidator {
         }
     }
 
+    fn get_validator_set(&self, height: u64) -> SnapchainValidatorSet {
+        let mut result = &self.validator_set[0];
+        for config in &self.validator_set {
+            if config.shard_ids.contains(&self.shard_id)
+                && config.effective_at <= height
+                && config.effective_at > result.effective_at
```

Comment:
This logic is duplicated in the shard validator as well. I think it might be good to push this into the `StoredValidatorSet` (or create another struct that holds it) so it stores the vector inside, and we can just ask it for the validator set for a height

Reply:
Done!